graph/tags.go

```diff
@@ -434,54 +437,98 @@ func validateDigest(dgst string) error {
 	return nil
 }
 
-// poolAdd checks if a push or pull is already running, and returns
-// (broadcaster, true) if a running operation is found. Otherwise, it creates a
-// new one and returns (broadcaster, false).
-func (store *TagStore) poolAdd(kind, key string) (*broadcaster.Buffered, bool) {
+// acquirePull checks if a pull is already running, and returns the broadcaster
+// associated with the existing pull and found = true if a running pull is
+// found. Otherwise, it creates a new one and returns the broadcaster and
+// found = false.
+//
+// If a push in the same repository is running, acquirePull waits for the push
+// to finish before finding or creating the broadcaster for the pull. waitFunc
+// is called to provide the caller the opportunity to print a message explaining
+// the delay.
+//
+// releasePull should only be called in cases where acquirePull returned
+// found = true.
+func (store *TagStore) acquirePull(key, repo string, waitFunc func()) (b *broadcaster.Buffered, found bool) {
 	store.Lock()
 	defer store.Unlock()
 
-	if p, exists := store.pullingPool[key]; exists {
-		return p, true
+	waitFuncCalled := false
+	for {
+		if store.pushCountByRepo[repo] > 0 {
+			if waitFunc != nil && !waitFuncCalled {
+				store.Unlock()
+				waitFunc()
+				waitFuncCalled = true
+				store.Lock()
+			} else {
+				store.pushPullCond.Wait()
+			}
+		} else {
+			break
+		}
 	}
-	if p, exists := store.pushingPool[key]; exists {
+
+	if p, exists := store.pullsByKey[key]; exists {
 		return p, true
 	}
 
-	broadcaster := broadcaster.NewBuffered()
+	b = broadcaster.NewBuffered()
+	store.pullsByKey[key] = b
+	store.pullCountByRepo[repo]++
+	return b, false
+}
+
+// acquirePush waits for all pulls and pushes associated with the repository
+// to be done.
+//
+// If a push or pull in the same repository is running, acquirePull waits for
+// the pull or push to finish. waitFunc is called to provide the caller the
+// opportunity to print a message explaining the delay.
+//
+// releasePush needs to be called after the push is finished.
+func (store *TagStore) acquirePush(repo string, waitFunc func()) {
```

Comment:
> Something else I'm wondering about, does repo string contain the tag? Seems like we would want to do things at the tag level so say, pushing foo:bar doesn't block pushing foo:baz

The reason it's done at the repo level is so that `docker push -a myrepo` and `docker push myrepo:latest` will wait for each other.

It would be possible to make this finer-grained. The locking/waiting would have to happen the level of the individual tag instead of the whole push. Managing the interaction with pulls would be more complex if we did it this way, since currently pulls don't register themselves on a per-tag basis. But if we decide that we don't care about pulling at the same time as pushing, it would be pretty straightforward.